File: src/main/java/com/tmvkrpxl0/tcombat/client/renderers/BlockEntityRenderer.java
package com.tmvkrpxl0.tcombat.client.renderers;

import com.mojang.blaze3d.matrix.MatrixStack;
import com.tmvkrpxl0.tcombat.common.entities.misc.CustomizableBlockEntity;
import net.minecraft.block.BlockRenderType;
import net.minecraft.block.BlockState;
import net.minecraft.client.Minecraft;
import net.minecraft.client.renderer.BlockRendererDispatcher;
import net.minecraft.client.renderer.IRenderTypeBuffer;
import net.minecraft.client.renderer.RenderTypeLookup;
import net.minecraft.client.renderer.entity.EntityRenderer;
import net.minecraft.client.renderer.entity.EntityRendererManager;
import net.minecraft.client.renderer.model.IBakedModel;
import net.minecraft.client.renderer.texture.AtlasTexture;
import net.minecraft.client.renderer.texture.OverlayTexture;
import net.minecraft.util.ResourceLocation;
import net.minecraft.util.math.BlockPos;
import net.minecraft.world.World;
import org.lwjgl.opengl.GL11;

import javax.annotation.Nonnull;
import java.util.Random;

public class BlockEntityRenderer  extends EntityRenderer<CustomizableBlockEntity> {

    public BlockEntityRenderer(EntityRendererManager renderManager) {
        super(renderManager);
        this.shadowRadius = 0.5F;
    }

    @Override
    public void render(CustomizableBlockEntity entityIn, float entityYaw, float partialTicks, @Nonnull MatrixStack matrixStackIn, @Nonnull IRenderTypeBuffer bufferIn, int packedLightIn) {
        BlockState blockstate = entityIn.getBlockState();
        if (blockstate.getRenderShape() == BlockRenderType.MODEL) {
            World world = entityIn.getCommandSenderWorld();
            if (blockstate != world.getBlockState(entityIn.blockPosition()) && blockstate.getRenderShape() != BlockRenderType.INVISIBLE) {
                matrixStackIn.pushPose();
                BlockPos blockpos = new BlockPos(entityIn.getX(), entityIn.getBoundingBox().maxY, entityIn.getZ());
                matrixStackIn.translate(-0.5D, 0.0D, -0.5D);
                BlockRendererDispatcher blockrendererdispatcher = Minecraft.getInstance().getBlockRenderer();
                IBakedModel test = blockrendererdispatcher.getBlockModel(blockstate);
                for (net.minecraft.client.renderer.RenderType type : net.minecraft.client.renderer.RenderType.chunkBufferLayers()) {
                    if (RenderTypeLookup.canRenderInLayer(blockstate, type)) {
                        net.minecraftforge.client.ForgeHooksClient.setRenderLayer(type);
                        blockrendererdispatcher.getModelRenderer().tesselateBlock(world, blockrendererdispatcher.getBlockModel(blockstate), blockstate, blockpos, matrixStackIn, bufferIn.getBuffer(type), false, new Random(), 2340965, OverlayTexture.NO_OVERLAY);
                    }
                }
                net.minecraftforge.client.ForgeHooksClient.setRenderLayer(null);
                matrixStackIn.popPose();
                super.render(entityIn, entityYaw, partialTicks, matrixStackIn, bufferIn, packedLightIn);
            }
        }
    }

    @Nonnull
    @Override
    public ResourceLocation getTextureLocation(@Nonnull CustomizableBlockEntity entity) {
        return AtlasTexture.LOCATION_BLOCKS;
    }
}

File: src/main/kotlin/com/tmvkrpxl0/tcombat/client/renderers/BlockEntityRenderer.kt
package com.tmvkrpxl0.tcombat.client.renderers

import com.mojang.blaze3d.matrix.MatrixStack
import com.tmvkrpxl0.tcombat.common.entities.misc.CustomizableBlockEntity
import net.minecraft.block.BlockRenderType
import net.minecraft.client.Minecraft
import net.minecraft.client.renderer.IRenderTypeBuffer
import net.minecraft.client.renderer.RenderType
import net.minecraft.client.renderer.RenderTypeLookup
import net.minecraft.client.renderer.entity.EntityRenderer
import net.minecraft.client.renderer.entity.EntityRendererManager
import net.minecraft.client.renderer.texture.AtlasTexture
import net.minecraft.client.renderer.texture.OverlayTexture
import net.minecraft.util.ResourceLocation
import net.minecraft.util.math.BlockPos
import net.minecraftforge.client.ForgeHooksClient
import java.util.*
import javax.annotation.Nonnull

class BlockEntityRenderer(renderManager: EntityRendererManager) : EntityRenderer<CustomizableBlockEntity>(renderManager) {
    init {
        this.shadowRadius = 0.5f
    }

    override fun render(entityIn: CustomizableBlockEntity, entityYaw: Float, partialTicks: Float, @Nonnull matrixStackIn: MatrixStack, @Nonnull bufferIn: IRenderTypeBuffer, packedLightIn: Int) {
        val blockState = entityIn.getBlockState()
        if (blockState.renderShape == BlockRenderType.MODEL) {
            val world = entityIn.level
            matrixStackIn.pushPose()
            val blockPos = BlockPos(entityIn.x, entityIn.boundingBox.maxY, entityIn.z)
            matrixStackIn.translate(-0.5, 0.0, -0.5)
            val blockRendererDispatcher = Minecraft.getInstance().blockRenderer
            for (type in RenderType.chunkBufferLayers()) {
                if (RenderTypeLookup.canRenderInLayer(blockState, type)) {
                    ForgeHooksClient.setRenderLayer(type)
                    blockRendererDispatcher.modelRenderer.tesselateBlock(
                        world, blockRendererDispatcher.getBlockModel(blockState), blockState, blockPos, matrixStackIn, bufferIn.getBuffer(type), false, Random(), 2340965, OverlayTexture.NO_OVERLAY
                    )
                }
            }
            ForgeHooksClient.setRenderLayer(null)
            matrixStackIn.popPose()
            super.render(entityIn, entityYaw, partialTicks, matrixStackIn, bufferIn, packedLightIn)
        }
    }

    @Nonnull
    override fun getTextureLocation(@Nonnull entity: CustomizableBlockEntity): ResourceLocation {
        return AtlasTexture.LOCATION_BLOCKS
    }
}